src/lib.rs

```diff
@@ -14,256 +14,321 @@
+                // translatable group starts here.
+                if let State::Skip(start) = state {
+                    groups.push(Group::Skip(&events[start..idx]));
+                    state = State::Translate(idx);
+                }
+            }
+            _ => {
+                // If we're currently translating, then a new
+                // skippable group starts here.
+                if let State::Translate(start) = state {
+                    groups.push(Group::Translate(&events[start..idx]));
+                    state = State::Skip(idx);
+                }
             }
         }
-
-        self.msgs
-            .push(Message::new(self.line_number(span.start), span));
-        self.message_open = true;
     }
 
-    /// Calculate the line number for the given offset.
-    fn line_number(&self, offset: usize) -> usize {
-        self.offsets.partition_point(|&o| o < offset) + 1
+    match state {
+        State::Translate(start) => groups.push(Group::Translate(&events[start..])),
+        State::Skip(start) => groups.push(Group::Skip(&events[start..])),
     }
 
-    /// Push a new Markdown event into the accumulator.
-    fn push_event(&mut self, evt: Event<'a>, span: Range<usize>) {
-        #[cfg(test)]
-        println!("{evt:?} -- {:?}", &self.document[span.start..span.end]);
-
-        // Track the nesting depth.
-        match evt {
-            Event::Start(_) => self.depth += 1,
-            Event::End(_) => self.depth -= 1,
-            _ => {}
-        }
+    groups
+}
 
-        // Handle skip_until_depth, including skipping the End event
-        // that returned to the desired level.
-        if let Some(depth) = self.skip_until_depth {
-            if self.depth <= depth {
-                self.skip_until_depth = None;
-            }
-            return;
-        }
+/// Render a slice of Markdown events back to Markdown.
+///
+/// # Examples
+///
+/// ```
+/// use mdbook_i18n_helpers::{extract_events, reconstruct_markdown};
+/// use pulldown_cmark::{Event, Tag};
+///
+/// let group = extract_events("Hello *world!*", None);
+/// let (reconstructed, _) = reconstruct_markdown(&group, None);
+/// assert_eq!(reconstructed, "Hello _world!_");
+/// ```
+///
+/// Notice how this will normalize the Markdown to use `_` for
+/// emphasis and `**` for strong emphasis. The style is chosen to
+/// match the [Google developer documentation style
+/// guide](https://developers.google.com/style/text-formatting).
+pub fn reconstruct_markdown(
+    group: &[(usize, Event)],
+    state: Option<State<'static>>,
+) -> (String, State<'static>) {
+    let events = group.iter().map(|(_, event)| event);
+    let mut markdown = String::new();
+    let options = Options {
+        code_block_token_count: 3,
+        list_token: '-',
+        emphasis_token: '_',
+        strong_token: "**",
+        ..Options::default()
+    };
+    let new_state = cmark_resume_with_options(
```

Comment:
```suggestion
    // Advance the true state, but throw away the rendered Markdown since it can contain padding.
    let new_state = cmark_resume_with_options(
```